### PR TITLE
feat(encoding):  implement private and public key exchange format

### LINKED ISSUE #78: feat(encoding): implement private and public key exchange format (slip32/bech32 serialization)
The private and public key serialization format is succintly described in [wip-adansdpc-hdwallets], reproduced here:

### Serialization format
    
Extended public and private keys are serialized as follows:

| Size     | Description |
|----------|-------------|
| 1 byte   | Depth: `0x00` for master nodes, `0x01` for level-1 derived keys, etc. |
| 4*depth bytes | Serialized path; each entry is encoded as 32-bit unsigned integer, big endian |
| 32 bytes | The chain code |
| 33 bytes | The public key or private key data |

The key data field will be ser<sub>P</sub>(K) for public keys and `0x00` || ser<sub>256</sub>(k) for private keys. 

This structure can be encoded using Bech32 format described in [WIP-adansdpc-addresses](). We will use 'xpub' human-readable part for extended public keys and 'xprv' for extended private keys.

When importing a serialized extended public key, implementations must verify whether the `X` coordinate in the public key data corresponds to a point on the curve. If not, the extended public key is invalid.

[wip-adansdpc-hdwallets]: https://github.com/aesedepece/WIPs/blob/wip-adansdpc-hdwallets/wip-adansdpc-hdwallets.md

### PR DESCRIPTION
**Pull request by mariocao**
# PR Prelude

Thank you for contributing to sheikah! :)

**Please fill in this template (by putting an `x` inside
the brackets) _before_ filing your PR:**

- [x] I have read and understood [CODE_OF_CONDUCT][code] document.
- [x] I have read and understood [CONTRIBUTING][cont] document.
- [x] I have read and understood [docs/styleguide][style] document.
- [x] I have included tests for the changes in my PR. If not, I have included a
  rationale for why I haven't.
- [x] My code is formatted and is accepted by `yarn fmt-verify`.
- [x] **I understand my PR may be closed if it becomes obvious I didn't
  actually perform all of these steps.**

# Why this change is necessary and useful
It implements private and public key exchange capabilities using slip32 and bech32 serialization.

Closes #78 

[code]: https://github.com/witnet/sheikah/blob/master/docs/CODE_OF_CONDUCT.md
[cont]: https://github.com/witnet/sheikah/blob/master/docs/CONTRIBUTING.md
[style]: https://github.com/witnet/sheikah/blob/master/docs/STYLEGUIDE.md

### CODE REVIEW

File: app/lib/encoding/slip32.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * Import key from slip32 format
+ * @param {string} slip32
+ * @returns {{keyPath: KeyPath; extendedKey: ExtendedKey<PrivateKey> | ExtendedKey<PublicKey>}}
+ */
+export const importKeyFromSlip32 = (slip32: string): {
+  keyPath: KeyPath.KeyPath,
+  extendedKey: Key.ExtendedKey<PrivateKey.PrivateKey> | Key.ExtendedKey<PublicKey.PublicKey>
+} => {
+  // Decode slip32
+  const {prefix, words}: { prefix: string, words: Buffer } = decode(slip32)
```
Comment (aesedepece) [inline app/lib/encoding/slip32.ts]: - Nice es6 destructuring!
- Here `prefix` should be `hrp` too, as that's the naming [SLIP-0032][slip] itself uses.
- This is also a `DecodedBech32`

[slip]: https://github.com/satoshilabs/slips/blob/master/slip-0032.md#serialization-format

File: app/lib/encoding/slip32.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  // Convert from words to bytes
+  const data = Bech32.fromWords(words)
+
+  // Extract Key Path depth
+  const depth: number = Buffer.from(data).readIntBE(0, DEPTH_LENGTH)
+
+  // Check data length
+  if (data.length !== DEPTH_LENGTH + depth * KEYPATH_LENGTH + CHAINCODE_LENGTH + KEY_LENGTH) {
+    throw Error("Malformed slip32 serialized key: invalid data length" +
+      `(expected: ${DEPTH_LENGTH + depth * KEYPATH_LENGTH + CHAINCODE_LENGTH + KEY_LENGTH},` +
```
Comment (aesedepece) [inline app/lib/encoding/slip32.ts]: Here's some code repetition / computation redundancy.

File: app/lib/encoding/slip32.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  // Convert from words to bytes
+  const data = Bech32.fromWords(words)
+
+  // Extract Key Path depth
+  const depth: number = Buffer.from(data).readIntBE(0, DEPTH_LENGTH)
+
+  // Check data length
+  if (data.length !== DEPTH_LENGTH + depth * KEYPATH_LENGTH + CHAINCODE_LENGTH + KEY_LENGTH) {
```
Comment (aesedepece) [inline app/lib/encoding/slip32.ts]: If we move this computation into its own function, we can have `depth` as an argument and achive better clarity here.

File: app/lib/encoding/slip32.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+const BECH32_LIMIT = DEPTH_LENGTH + ((Math.pow(2, DEPTH_LENGTH * 8) - 1) * KEYPATH_LENGTH)
+  + CHAINCODE_LENGTH + KEY_LENGTH
+
+/**
+ * Bech32 decode wrapper using the predefined limit
+ * @param {string} string
+ * @returns {{prefix: string; words: Buffer}}
+ *         Human readable part and decoded words (5-bit)
+ */
+export const decode = (string: string): { prefix: string, words: Buffer } => {
```
[the remark this answers is not in the document]
Comment (mariocao) [reply app/lib/encoding/slip32.ts]: As discussed with @anler and @aesedepece , maybe *PrefixedBuffer* may be also a good candidate. :)

File: test/app/lib/encoding/slip32.spec.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
+import * as exchange from "../../../../app/lib/encoding/slip32"
+import * as KeyPath from "../../../../app/lib/crypto/keyPath"
+import * as PrivateKey from "../../../../app/lib/crypto/key/privateKey"
+import * as PublicKey from "../../../../app/lib/crypto/key/publicKey"
+import * as Key from "../../../../app/lib/crypto/key/key"
+
+import * as fixtures from "./fixtures"
```
Comment (aesedepece) [inline test/app/lib/encoding/slip32.spec.ts]: Love the idea of putting invariants into a separate file :clap:

File: test/app/lib/encoding/slip32.spec.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  const buffer = Buffer.from(hex, "hex")
+  const key = buffer.slice(buffer.length - 33)
+  const chaincode = buffer.slice(buffer.length - 33 - 32, buffer.length - 33)
+
+  return {key, chaincode}
+}
+
+/**
+ * Crop long strings to be displayed in the tests
+ * @param {string} text
```
Comment (aesedepece) [inline test/app/lib/encoding/slip32.spec.ts]: Signature in docs doesn't match actual signature in implementation.

File: test/app/lib/encoding/slip32.spec.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+ * @returns {string}
+ */
+function cropString(text: string, limit: number, end?: boolean) {
+  if (!end) {
```
Comment (aesedepece) [inline test/app/lib/encoding/slip32.spec.ts]: `end` could use an explanation on what is it for. AFAICT, it tells the function to crop by the beginning (default) or end of the string, but that's not so obvious from the code itself :sweat_smile:

File: app/lib/encoding/slip32.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+  return Bech32.decode(string, BECH32_LIMIT)
+}
+
+/**
+ * Bech32 decode wrapper using the predefined limit
+ * @param {string} hrp
+ * @param {Buffer} words
+ * @returns {string}
+ */
+export const encode = (hrp: string, words: Buffer): string => {
```
[the remark this answers is not in the document]
Comment (mariocao) [reply app/lib/encoding/slip32.ts]: Is it enough to modify the signature of encode as follows?
```
export const encode = ({hrp, words}: PrefixedBuffer): string => {
```

[the remark this answers is not in the document]
Comment (aesedepece) [reply app/lib/encoding/slip32.ts]: Yay, that's cool. Argument destructuring at its best :+1:

File: app/lib/encoding/slip32.ts [hunk 9]
```diff
@@ -79,9 +88,9 @@ export const importKeyFromSlip32 = (slip32: string): {
   extendedKey: Key.ExtendedKey<PrivateKey.PrivateKey> | Key.ExtendedKey<PublicKey.PublicKey>
 } => {
   // Decode slip32
-  const {prefix, words}: { prefix: string, words: Buffer } = decode(slip32)
+  const {hrp, words}: { hrp: string, words: Buffer } = decode(slip32)
   // Check hrp
-  if ((prefix !== "xprv") && (prefix !== "xpub")) {
+  if ((hrp !== "xprv") && (hrp !== "xpub")) {
```
Comment (aesedepece) [inline app/lib/encoding/slip32.ts]: Side comment (no changes required here): on the long term, we could define if we favor conjunction of negation `a !== b && a !== c` or negation of disjunction `!(a == b || a == c)` unsing [De Morgan's laws][demorgan] to improve readability of some logic assertions with tricky semantics.

In this case, I think it's just OK as is (we save one operator).

This is also equivalent to `!["xpriv", "xpub"].includes(hrp)`, which can be cool when you need to match against a lot of different invariants, but it has a serious performance price.

[demorgan]: https://en.wikipedia.org/wiki/De_Morgan%27s_laws

**Comment by aesedepece** [comment]
This can be merged once `yarn.lock` conflict is solved.

**Pull request merged**